lib/ret_web/controllers/api/v1/avatar_controller.ex

```diff
@@ -0,0 +1,192 @@
+defmodule RetWeb.Api.V1.AvatarController do
+  use RetWeb, :controller
+
+  alias Ret.{Account, Repo, Avatar, Storage, OwnedFile}
+
+  plug(RetWeb.Plugs.RateLimit when action in [:create, :update])
+
+  @primary_material_name "Bot_PBS"
+  @texture_paths %{
+    base_map_owned_file: [["pbrMetallicRoughness", "baseColorTexture"]],
+    emissive_map_owned_file: [["emissiveTexture"]],
+    normal_map_owned_file: [["normalTexture"]],
+    orm_map_owned_file: [
+      ["pbrMetallicRoughness", "metallicRoughnessTexture"],
+      ["occlusionTexture"]
+    ]
+  }
+  @image_columns Map.keys(@texture_paths)
+  @file_columns [:gltf_owned_file, :bin_owned_file] ++ @image_columns
+
+  def create(conn, %{"avatar" => params}) do
+    create_or_update(conn, params, %Avatar{})
+  end
+
+  def update(conn, %{"id" => avatar_sid, "avatar" => params}) do
+    case avatar_sid |> get_avatar() do
+      %Avatar{} = avatar -> create_or_update(conn, params, avatar)
+      _ -> conn |> send_resp(404, "not found")
+    end
+  end
+
+  defp create_or_update(
+         conn,
+         _params,
+         %Avatar{account_id: avatar_account_id},
+         %Account{account_id: account_id}
+       )
+       when not is_nil(avatar_account_id) and avatar_account_id != account_id do
+    conn |> send_resp(401, "")
+  end
+
+  defp create_or_update(conn, params, avatar) do
+    account = conn |> Guardian.Plug.current_resource()
+
+    files_to_promotoe =
+      case params["files"] do
+        nil ->
+          %{}
+
+        files ->
+          files
+          |> Enum.map(fn {k, v} -> {String.to_atom(k), List.to_tuple(v)} end)
+          |> Enum.into(%{})
+      end
+
+    owned_file_results = Storage.promote(files_to_promotoe, account)
+
+    promotion_error =
+      owned_file_results |> Map.values() |> Enum.filter(&(elem(&1, 0) == :error)) |> Enum.at(0)
```

Comment:
it may be worth revisiting the return value of the map variant of `promote` to return a wrapper tuple that encodes if there is any error, so this weird destructuring can go away here and in scene controller